osu.Game/Beatmaps/WorkingBeatmap.cs

```diff
@@ -114,11 +115,57 @@ private void applyRateAdjustments(Track t = null)
                 mod.ApplyToClock(t);
         }
 
-        public class AsyncLazy<T> : Lazy<Task<T>>
+        public class AsyncLazy<T>
         {
-            public AsyncLazy(Func<T> valueFactory)
-                : base(() => Task.Run(valueFactory))
+            private Lazy<Task<T>> lazy;
+            private readonly Func<T> valueFactory;
+            private readonly Func<T, bool> stillValidFunction;
+
+            public AsyncLazy(Func<T> valueFactory, Func<T, bool> stillValidFunction = null)
+            {
+                this.valueFactory = valueFactory;
+                this.stillValidFunction = stillValidFunction;
+
+                init();
+            }
+
+            public void Recycle()
+            {
+                if (!IsValueCreated) return;
+
+                (lazy.Value.Result as IDisposable)?.Dispose();
+
+                init();
+            }
+
+            public bool IsValueCreated
+            {
+                get
+                {
+                    ensureValid();
+                    return lazy.IsValueCreated;
+                }
+            }
+
+            public Task<T> Value
+            {
+                get
+                {
+                    ensureValid();
+                    return lazy.Value;

```

Comment:
You realise that doing this removes any thread-safety guarantees Lazy.Value previously provided, right?

Reply:
how so?

Reply:
If two threads enter Value -> init() at the same time, then one thread continues on to return lazy.Value, and the other thread does the same afterwards, the object will be initialised twice.

Reply:
fair call. will add some local locking.